external-crates/move/crates/move-vm-runtime/src/tracing2/tracer.rs

```diff
@@ -0,0 +1,1812 @@
+mod vm_tracer_noop {
+    use move_trace_format::format::MoveTraceBuilder;
+
+    #[allow(dead_code)]
+    pub(crate) struct VMTracer<'a> {
+        _phantom: std::marker::PhantomData<&'a ()>,
+    }
+
+    impl<'a> VMTracer<'a> {
+        #[allow(dead_code)]
+        pub(crate) fn new(_trace: &'a mut MoveTraceBuilder) -> Self {
+            Self {
+                _phantom: std::marker::PhantomData,
+            }
+        }
+    }
+}
+
+#[cfg(feature = "gas-profiler")]
+pub(crate) use vm_tracer_impl::*;
+#[cfg(feature = "gas-profiler")]
+mod vm_tracer_impl {
+    use crate::{
+        interpreter::{Frame, Interpreter},
+        loader::{Function, Loader},
+    };
+    use move_binary_format::{
+        errors::PartialVMError,
+        file_format::{ConstantPoolIndex, SignatureIndex},
+    };
+    use move_core_types::{
+        account_address::AccountAddress,
+        annotated_value::{MoveTypeLayout, MoveValue},
+        language_storage::TypeTag,
+    };
+    use move_trace_format::format::{
+        DataLoad, Effect as EF, Location, MoveTraceBuilder, Read, RefType, TraceIndex, TraceValue,
+        TypeTagWithRefs, Write,
+    };
+    use move_vm_types::{loaded_data::runtime_types::Type, values::Value};
+    use std::collections::BTreeMap;
+
+    /// Internal state for the tracer. This is where the actual tracing logic is implemented.
+    pub(crate) struct VMTracer<'a> {
+        trace: &'a mut MoveTraceBuilder,
+        link_context: Option<AccountAddress>,
+        pc: Option<u16>,
+        active_frames: BTreeMap<TraceIndex, FrameInfo>,
+        type_stack: Vec<RootedType>,
+        loaded_data: BTreeMap<TraceIndex, TraceValue>,
+        effects: Vec<EF>,
+    }
+
+    /// Information about a frame that we keep during trace building
+    #[derive(Debug, Clone)]
+    struct FrameInfo {
+        is_native: bool,
+        locals_types: Vec<LocalType>,
+        return_types: Vec<TagWithLayoutInfoOpt>,
+    }
+
+    /// A type tag, and the move type layout and reference information for that type if it is
+    /// computable without error. Due to runtime value depth restrictions you can have a valid type
+    /// whose type layout is not computable at runtime without error.
+    #[derive(Debug, Clone)]
+    struct TagWithLayoutInfoOpt {
+        tag: TypeTag,
+        layout: (Option<MoveTypeLayout>, Option<RefType>),
+    }
+
+    // Information about a function that we use for trace building
+    // All types are fully substituted
+    #[derive(Debug, Clone)]
+    struct FunctionTypeInfo {
+        ty_args: Vec<TypeTag>,
+        local_types: Vec<TagWithLayoutInfoOpt>,
+        return_types: Vec<TagWithLayoutInfoOpt>,
+    }
+
+    /// A runtime location can refer to the stack to make it easier to refer to values on the stack and
+    /// resolving them. However, the stack is not a valid location for a reference and all references
+    /// are rooted in a local so the Trace `Location` does not include the stack, and only `Local` and
```

Comment:
I think this is not the main point of this comment, but can't references also be rooted in `Global`s?

Reply:
They can! Will update